beacon-chain/p2p/handshake.go

```diff
@@ -4,51 +4,134 @@ import (
-			s.peers.Add(nil /* ENR */, conn.RemotePeer(), conn.RemoteMultiaddr(), conn.Stat().Direction)
-			if len(s.peers.Active()) >= int(s.cfg.MaxPeers) {
-				go func() {
+			// Connection handler must be non-blocking as part of libp2p design.
+			go func() {
+				if peerHandshaking(remotePeer) {
+					// Exit this if there is already another connection
+					// request in flight.
+					return
+				}
+				defer peerFinished(remotePeer)
+				// Handle the various pre-existing conditions that will result in us not handshaking.
+				peerConnectionState, err := s.peers.ConnectionState(remotePeer)
+				if err == nil && (peerConnectionState == peers.PeerConnected || peerConnectionState == peers.PeerConnecting) {
+					log.WithField("currentState", peerConnectionState).WithField("reason", "already active").Trace("Ignoring connection request")
+					return
+				}
+				s.peers.Add(nil /* ENR */, remotePeer, conn.RemoteMultiaddr(), conn.Stat().Direction)
+				if len(s.peers.Active()) >= int(s.cfg.MaxPeers) {
 					log.WithField("reason", "at peer limit").Trace("Ignoring connection request")
-					if err := goodbyeFunc(context.Background(), conn.RemotePeer()); err != nil {
+					if err := goodbyeFunc(context.Background(), remotePeer); err != nil {
 						log.WithError(err).Trace("Unable to send goodbye message to peer")
 					}
-					if err := s.Disconnect(conn.RemotePeer()); err != nil {
-						log.WithError(err).Error("Unable to disconnect from peer")
+					// Add a short delay to allow the stream to flush before closing the connection.
+					// There is still a chance that the peer won't receive the message.
+					time.Sleep(flushDuration)
+					disconnectFromPeer()
+					return
+				}
+				if s.peers.IsBad(remotePeer) {
+					log.WithField("reason", "bad peer").Trace("Ignoring connection request")
+					disconnectFromPeer()
+					return
+				}
+
+				// Do not perform handshake on inbound dials.
+				if conn.Stat().Direction == network.DirInbound {
+					_, err := s.peers.ChainState(remotePeer)
+					peerExists := err == nil
+					currentTime := roughtime.Now()
+
+					// Wait for peer to initiate handshake
+					time.Sleep(timeForStatus)
+
+					// Exit if we are disconnected with the peer.
+					if s.host.Network().Connectedness(remotePeer) != network.Connected {
+						return
 					}
-				}()
-				return
-			}
-			if s.peers.IsBad(conn.RemotePeer()) {
-				log.WithField("reason", "bad peer").Trace("Ignoring connection request")
-				if err := s.Disconnect(conn.RemotePeer()); err != nil {
-					log.WithError(err).Error("Unable to disconnect from peer")
+
+					// If peer hasn't sent a status request, we disconnect with them
+					if _, err := s.peers.ChainState(remotePeer); err == peers.ErrPeerUnknown {
+						disconnectFromPeer()
+						return
+					}
+					if peerExists {
+						updated, err := s.peers.ChainStateLastUpdated(remotePeer)
+						if err != nil {
+							disconnectFromPeer()
+							return
+						}
+						if !updated.After(currentTime) {
```

Comment:
why are we disconnecting in this case?

Reply:
Because they haven't sent a status message.

Reply:
thanks